pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,308 @@
+				To(Equal(kubefedopt.DefaultFederatedVersion))
+		})
+
+		It("the returned resource's name should be the same as the input's object", func() {
+			Expect(expectedTyped.GetName()).
+				To(Equal(resourceTyped.GetName()))
+		})
+
+		It("the returned resource's template should be the input's unstructured", func() {
+			// This test implicitly tests that the nested hierarchy of fields exists
+
+			mapObj, found, err := getTemplateField(expectedTyped)
+			targetResource := &unstructured.Unstructured{}
+			// No need to check error code because it's covered by another test
+			_ = f.scheme.Convert(resourceTyped, targetResource, nil)
+			removeUnwantedFields(targetResource)
+
+			Expect(err).ToNot(HaveOccurred())
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+			Expect(mapObj["spec"]).To(BeEquivalentTo(targetResource.Object["spec"]))
+			Expect(mapObj["metadata"]).To(BeEquivalentTo(targetResource.Object["metadata"]))
+		})
+
+		It("the returned resource's placement should be an empty matcher", func() {
+			// This test implicitly tests that the nested hierarchy of fields exists
+			mapObj, found, err := getMatchLabelsField(expectedTyped)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+			Expect(mapObj).To(BeEmpty())
+		})
+
+		It("the returned resource's namespace should be the input's namespace", func() {
+			Expect(expectedTyped.GetNamespace()).
+				To(Equal(resourceTyped.GetNamespace()))
+		})
+	})
+}
+
+// func assertNestedFields(accessor interface{}, expectedObj types.GomegaMatcher, obj map[string]interface{}, fields ...string) { }
+func testDistribute() {
+	var (
+		resourceTyped       *corev1.Pod
+		fedResource, fedPod *unstructured.Unstructured
+		f                   *federator
+		err                 error
+		initObjs            []runtime.Object
+		clusterNames        []string
+		scheme              *runtime.Scheme
+	)
+
+	BeforeEach(func() {
+		resourceTyped = newPod("test-pod")
+		scheme = runtime.NewScheme()
+	})
+
+	JustBeforeEach(func() {
+		f = newFederatorWithScheme(scheme, initObjs...)
+		err = f.Distribute(resourceTyped, clusterNames...)
+	})
+
+	When("the resource type scheme has NOT been added to the type registry", func() {
+		It("should throw an error", func() {
+			Expect(err).To(HaveOccurred())
+		})
+	})
+
+	When("the resource type scheme has been added to the type registry", func() {
+		BeforeEach(func() {
+			corev1.AddToScheme(scheme)
+			tmpFederator := newFederatorWithScheme(scheme)
+			fedResource, _ = tmpFederator.createFederatedResource(resourceTyped)
+		})
+
+		JustBeforeEach(func() {
+			fedPod = getResourceFromAPI(f.kubeFedClient, fedResource)
+		})
+
+		It("should return no error on success", func() {
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		When("the resource is already in the Kube API", func() {
+			expectedImage := "apache"
+
+			BeforeEach(func() {
+				initObjs = append(initObjs, fedResource)
```

Comment:
Instead of exposing internals here, you can have the tests use Distribute to create the original, and just test that a second call to Distribute actually updates it.

Reply:
Because the test is for testing Distribute and I find it logically incorrect to do so.